flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/math/distributed/DistributedRowMatrix.scala

```diff
@@ -0,0 +1,182 @@
+
+  /**
+    * Apply a high-order function to couple of rows
+    * @param fun
+    * @param other
+    * @return
+    */
+  def byRowOperation(fun: (Vector, Vector) => Vector,
+                     other: DistributedRowMatrix): DistributedRowMatrix = {
+    val otherData = other.getRowData
+    require(this.getNumCols == other.getNumCols)
+    require(this.getNumRows == other.getNumRows)
+
+
+    val result = this.data
+      .fullOuterJoin(otherData)
+      .where("rowIndex")
+      .equalTo("rowIndex")(
+          (left: IndexedRow, right: IndexedRow) => {
+            val row1 = Option(left).getOrElse(IndexedRow(
+                    right.rowIndex,
+                    SparseVector.fromCOO(right.values.size, List((0, 0.0)))))
+            val row2 = Option(right).getOrElse(IndexedRow(
+                    left.rowIndex,
+                    SparseVector.fromCOO(left.values.size, List((0, 0.0)))))
+
+            IndexedRow(row1.rowIndex, fun(row1.values, row2.values))
+          }
+      )
+    new DistributedRowMatrix(result, numRowsOpt, numColsOpt)
+  }
+
+  /**
+    * Add the matrix to another matrix.
+    * @param other
+    * @return
+    */
+  def sum(other: DistributedRowMatrix): DistributedRowMatrix = {
+    val sumFunction: (Vector, Vector) => Vector = (x: Vector, y: Vector) =>
+      (x.asBreeze + y.asBreeze).fromBreeze
+    this.byRowOperation(sumFunction, other)
+  }
+
+  /**
+    * Subtracts another matrix.
+    * @param other
+    * @return
+    */
+  def subtract(other: DistributedRowMatrix): DistributedRowMatrix = {
+    val subFunction: (Vector, Vector) => Vector = (x: Vector, y: Vector) =>
+      (x.asBreeze - y.asBreeze).fromBreeze
+    this.byRowOperation(subFunction, other)
+  }
+}
+
+object DistributedRowMatrix {
+
+  type MatrixRowIndex = Int
+
+  /**
+    * Builds a DistributedRowMatrix from a dataset in COO
+    * @param isSorted If false, sorts the row to properly build the matrix representation.
+    *                 If already sorted, set this parameter to true to skip sorting.
+    * @return
+    */
+  def fromCOO(data: DataSet[(Int, Int, Double)],
+              numRows: Int,
+              numCols: Int,
+              isSorted: Boolean = false): DistributedRowMatrix = {
+    val vectorData: DataSet[(Int, SparseVector)] = data
+      .groupBy(0)
+      .reduceGroup(sparseRow => {
+        require(sparseRow.nonEmpty)
+        val sortedRow =
+          if (isSorted) {
+            sparseRow.toList
+          } else {
+            sparseRow.toList.sortBy(row => row._2)
+          }
+        val (indices, values) = sortedRow.map(x => (x._2, x._3)).unzip
+        (sortedRow.head._1,
+         SparseVector(numCols, indices.toArray, values.toArray))
+      })
+
+    val zippedData = vectorData.map(x => IndexedRow(x._1.toInt, x._2))
+
+    new DistributedRowMatrix(zippedData, Some(numRows), Some(numCols))
+  }
+}
+
+case class IndexedRow(rowIndex: Int, values: Vector)
+    extends Ordered[IndexedRow] {
```

Comment:
I thought it was the most elegant and generic way to sort a list of IndexedRow. If it's not the suggested way in Flink's codebase, I can remove it.